File: backend/app/schemas/wrapped.py
"""Schemas para endpoints de wrapped anual."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class MemberInfo(BaseModel):
    user_id: str
    username: str
    display_name: str | None = None
    avatar_url: str | None = None


class MemberSuperlative(BaseModel):
    user_id: str
    username: str
    display_name: str | None = None
    avatar_url: str | None = None
    title: str
    emoji: str
    stat_label: str
    stat_value: str


class HighestRatedBook(BaseModel):
    title: str
    cover_url: str | None = None
    author: str | None = None
    avg_rating: float


class FunniestOneliner(BaseModel):
    text: str
    author_username: str
    author_display_name: str | None = None
    author_avatar_url: str | None = None
    vote_count: int


class MostEmotionalBook(BaseModel):
    title: str
    cover_url: str | None = None
    author: str | None = None
    cried_percentage: float


class GenreBreakdownItem(BaseModel):
    genre: str
    count: int
    percentage: float


class EmotionalStats(BaseModel):
    total_reviews: int
    cried_count: int
    loved_it_count: int
    felt_aroused_count: int
    found_heavy_count: int
    wants_more_count: int


class WrappedData(BaseModel):
    year: int
    group_name: str
    group_photo_url: str | None = None
    total_books_read: int
    total_pages: int
    total_reading_hours: float
    genre_breakdown: list[GenreBreakdownItem]
    highest_rated_book: HighestRatedBook | None = None
    most_active_member: MemberInfo | None = None
    longest_streak_member: MemberInfo | None = None
    funniest_oneliner: FunniestOneliner | None = None
    most_emotional_book: MostEmotionalBook | None = None
    member_superlatives: list[MemberSuperlative]
    emotional_stats: EmotionalStats
    member_avatars: list[MemberInfo]


class WrappedResponse(BaseModel):
    group_id: str
    year: int
    data: WrappedData
    generated_at: datetime
    generated_by: str

    model_config = {"from_attributes": True}

File: frontend/src/lib/types/wrapped.ts
export interface MemberInfo {
  user_id: string;
  username: string;
  display_name?: string | null;
  avatar_url?: string | null;
}

export interface MemberSuperlative {
  user_id: string;
  username: string;
  display_name?: string | null;
  avatar_url?: string | null;
  title: string;
  emoji: string;
  stat_label: string;
  stat_value: string;
}

export interface HighestRatedBook {
  title: string;
  cover_url?: string | null;
  author?: string | null;
  avg_rating: number;
}

export interface FunniestOneliner {
  text: string;
  author_username: string;
  author_display_name?: string | null;
  author_avatar_url?: string | null;
  vote_count: number;
}

export interface MostEmotionalBook {
  title: string;
  cover_url?: string | null;
  author?: string | null;
  cried_percentage: number;
}

export interface GenreBreakdownItem {
  genre: string;
  count: number;
  percentage: number;
}

export interface EmotionalStats {
  total_reviews: number;
  cried_count: number;
  loved_it_count: number;
  felt_aroused_count: number;
  found_heavy_count: number;
  wants_more_count: number;
}

export interface WrappedData {
  year: number;
  group_name: string;
  group_photo_url?: string | null;
  total_books_read: number;
  total_pages: number;
  total_reading_hours: number;
  genre_breakdown: GenreBreakdownItem[];
  highest_rated_book?: HighestRatedBook | null;
  most_active_member?: MemberInfo | null;
  longest_streak_member?: MemberInfo | null;
  funniest_oneliner?: FunniestOneliner | null;
  most_emotional_book?: MostEmotionalBook | null;
  member_superlatives: MemberSuperlative[];
  emotional_stats: EmotionalStats;
  member_avatars: MemberInfo[];
}

export interface WrappedResponse {
  group_id: string;
  year: number;
  data: WrappedData;
  generated_at: string;
  generated_by: string;
}
